### PR TITLE
Connector unit tests junit5/2

### PR DESCRIPTION
**Pull request by ebd2**
Updated per discussion in https://github.com/prestodb/presto/issues/8150
Depends on https://github.com/airlift/airbase/pull/110

Notable updates from https://github.com/prestodb/presto/pull/8672:

- Tests are now enabled/disabled based on supported/required features using SupportedTestCondition
- Cleanup of schemas and tables created for tests is more robust; a failed test is less likely to cause subsequent failed tests
- Junit5 and TestNG tests coexist (pending the airbase change)

### CODE REVIEW

**Comment by ebd2** [comment]
@electrum, @losipiuk, @kokosing

**Comment by ebd2** [comment]
Travis failure is expected. Code depends on unreleased airbase changes.

File: pom.xml [hunk 1]
```diff
@@ -5,7 +5,7 @@
     <parent>
         <groupId>io.airlift</groupId>
         <artifactId>airbase</artifactId>
-        <version>64</version>
+        <version>64-ERIC</version>
```
Comment (losipiuk) [inline pom.xml]: This commit adds `presto-connector-tests` maven module but not list that in `<modules>` section of top-level `pom.xml`.

File: presto-connector-tests/src/main/java/com/facebook/presto/connector/meta/FeatureUtil.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import com.google.common.collect.Sets;
+
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Stream;
+
+public class FeatureUtil
```
Comment (losipiuk) [inline presto-connector-tests/src/main/java/com/facebook/presto/connector/meta/FeatureUtil.java]: This class is later gone. Should it be introduced in the first place.

Also I tried to do the review commit by commit but it is hard as individual commits do not compile.
Please make sure that each commit compiles. Or at least does not have pom.xml errors which prevent IntelliJ to correctly parse project structure.

File: pom.xml [hunk 3]
```diff
@@ -935,6 +942,35 @@
...
+            <dependency>
+                <groupId>org.junit.jupiter</groupId>
+                <artifactId>junit-jupiter-api</artifactId>
+                <version>${dep.junit.jupiter.version}</version>
+            </dependency>
+
+            <dependency>
+                <groupId>junit</groupId>
+                <artifactId>junit</artifactId>
+                <version>${dep.junit.version}</version>
```
Comment (losipiuk) [inline pom.xml]: this is not longer part of airbase

Comment (ebd2) [inline pom.xml]: I updated airbase with David's feedback, but have not yet updated this PR.

**Comment by ebd2** [comment]
@losipiuk:

Commits squashed as requested. I'll update it again once I get the maven changes incorporated into the first commit.

**Comment by ebd2** [comment]
@losipiuk: Maven changes incorporated. Back to you.

File: presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/BaseMetadataTest.java [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    {
+        return new SchemaTableName(schemaName, tableName);
+    }
+
+    @Test
+    default void testEmptyMetadata()
+    {
+        ConnectorSession session = new TestingConnectorSession(ImmutableList.of());
+
+        run(ImmutableList.of(
```
Comment (losipiuk) [inline presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/BaseMetadataTest.java]: Or it is important that subsequent calls are done in separate transactions.

Still nicer looking approach would be to have `withMetadata` take multiple consumers:
```
 default void withMetadata(Consumer<ConnectorMetadata> ... consumer)
```

And then use:

```
        withMetadata(
                metadata -> assertEquals(metadata.listSchemaNames(session), systemSchemas()),
                metadata -> assertEquals(metadata.listTables(session, null), ImmutableList.of())
        );
```

File: presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/MetadataTableTest.java [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+        List<ColumnMetadata> columns = ImmutableList.of(
+                new ColumnMetadata("bigint_column", BIGINT),
+                new ColumnMetadata("double_column", DOUBLE));
+
+        List<ColumnMetadata> expectedColumns = withSystemColumns(columns);
+
+        ConnectorTableMetadata tableMetadata11 = new ConnectorTableMetadata(schemaTableName1, columns, getTableProperties());
+        ConnectorTableMetadata tableMetadata12 = new ConnectorTableMetadata(schemaTableName2, columns, getTableProperties());
+
+        try (AutoCloseable cleanup = withTables(session, tableMetadata11, tableMetadata12)) {
```
Comment (losipiuk) [inline presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/MetadataTableTest.java]: Maybe we should keep `API` consistent for `withMetadata`.
 this:

```java
    default void withTables(ConnectorSession session, List<ConnectorTableMetadata> tables, Runnable runnable)
            throws Exception
    {
        try (AutoCloseable cleanup = withTables(session, tables.toArray(new ConnectorTableMetadata[0]))) {
            runnable.run();
        }
    }

    default void withTable(
            ConnectorSession session,
            ConnectorTableMetadata table,
            Runnable runnable)
            throws Exception
    {
        withTables(session, Arrays.asList(table), runnable);
    }

    default void withTables(
            ConnectorSession session,
            ConnectorTableMetadata table1,
            ConnectorTableMetadata table2,
            Runnable runnable)
            throws Exception
    {
        withTables(session, Arrays.asList(table1, table2), runnable);
    }

    default AllCloser withTables(ConnectorSession session, ConnectorTableMetadata... tables)
    {
        List<ConnectorTableMetadata> tableList = Arrays.asList(tables);
        return new AllCloser(Streams.concat(
                withSchemas(session, schemaNamesOf(tableList)).stream(),
                tableList.stream()
                        .map(table -> new Table(this, session, table)))
                .collect(toImmutableList()));
    }
```

Then this code snippet (with changes to `withMetadata` suggested above) would look like:
```java
        withTables(session, tableMetadata11, tableMetadata12, () ->
                withMetadata(
                        metadata -> assertEquals(metadata.listTableColumns(session, new SchemaTablePrefix()),
                                ImmutableMap.of(
                                        schemaTableName1, expectedColumns,
                                        schemaTableName2, expectedColumns)),
                        metadata -> assertEquals(metadata.listTableColumns(session, prefixOf(schemaTableName1)),
                                ImmutableMap.of(
                                        schemaTableName1, expectedColumns))));
```

Nicer?

File: presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/BaseMetadataTest.java [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+        @Override
+        public void close()
+                throws Exception
+        {
+            test.withMetadata(metadata -> metadata.dropSchema(session, name));
+        }
+    }
+
+    default List<AutoCloseable> withSchemas(ConnectorSession session, List<String> schemaNames)
```
Comment (losipiuk) [inline presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/BaseMetadataTest.java]: same suggestion as for `withTables`. Is there a benefit from exposing `AutoClosable`s to caller. Are there cases when it would not be immediatelly wrapped in `try`. If that is not the case I suggest to use `Runnable` as last argument and use lambdas when it is used.

File: presto-connector-tests/src/main/java/com/facebook/presto/connector/meta/SupportedTestCondition.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.extension.ConditionEvaluationResult.disabled;
+import static org.junit.jupiter.api.extension.ConditionEvaluationResult.enabled;
+
+/*
+ * Determines whether a test method should run based on the @SupportedFeatures
+ * and @RequiredFeatures of the test method and the connector the test is being
+ * run against.
+ *
+ * Consider the following class hierarchy:
```
Comment (losipiuk) [inline presto-connector-tests/src/main/java/com/facebook/presto/connector/meta/SupportedTestCondition.java]: verbose Javadoc. As always :)

File: presto-tests/pom.xml [hunk 8]
```diff
@@ -195,7 +195,6 @@
         <dependency>
             <groupId>io.airlift</groupId>
             <artifactId>log-manager</artifactId>
-            <scope>test</scope>
```
Comment (losipiuk) [inline presto-tests/pom.xml]: unrelated?

File: pom.xml [hunk 9]
```diff
@@ -120,6 +120,7 @@
         <module>presto-thrift-testing-server</module>
         <module>presto-thrift-connector</module>
         <module>presto-matching</module>
+        <module>presto-connector-tests</module>
```
Comment (losipiuk) [inline pom.xml]: changes to root `pom.xml` should be part of `JUnit-5 unsupported test filtering by feature`. Right?

File: presto-connector-tests/pom.xml [hunk 10]
```diff
@@ -35,5 +35,11 @@
 			<groupId>org.junit.jupiter</groupId>
 			<artifactId>junit-jupiter-api</artifactId>
 		</dependency>
+
+		<dependency>
```
Comment (losipiuk) [inline presto-connector-tests/pom.xml]: indentation. Fix intellij settings.

File: presto-hive/src/test/java/com/facebook/presto/hive/unittests/TestHiveMetadata.java [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+    {
+        return ImmutableMap.of(
+                BUCKETED_BY_PROPERTY, ImmutableList.of(),
+                BUCKET_COUNT_PROPERTY, 0,
+                PARTITIONED_BY_PROPERTY, ImmutableList.of(),
+                STORAGE_FORMAT_PROPERTY, TEXTFILE);
+    }
+
+    @Override
+    public List<ColumnMetadata> withSystemColumns(List<ColumnMetadata> connectorColumns)
```
Comment (losipiuk) [inline presto-hive/src/test/java/com/facebook/presto/hive/unittests/TestHiveMetadata.java]: I had hard time guessing what this method does. 
Maybe call it `extendWithConnectorSpecificColumns`? Or just add `getConnectorSpecificColumns` method and to the concatenation of lists in tests?

File: presto-postgresql/src/test/java/com/facebook/presto/unittests/TestPostgresqlMetadata.java [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    @Override
+    public List<String> systemSchemas()
+    {
+        return ImmutableList.of("pg_catalog", "public");
+    }
+
+    @Override
+    public List<AutoCloseable> withSchemas(ConnectorSession session, List<String> schemaNames)
```
Comment (losipiuk) [inline presto-postgresql/src/test/java/com/facebook/presto/unittests/TestPostgresqlMetadata.java]: why are we overriding it? Aren't we supposed to filter out all the tests which require it by using `CREATE_TABLE_AS, DROP_TABLE` in header? Shouldn't only schema related tests use the `withSchemas` method.

File: presto-connector-tests/src/main/java/com/facebook/presto/connector/meta/SupportedTestCondition.java [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+
+        return requiredFeatures.isEmpty() ?
+                enabled("All required features present") :
+                disabled("Missing required features: " + Joiner.on(", ").join(requiredFeatures));
+    }
+
+    private static List<RequiredFeatures> getClassRequiredFeatures(Class<?> testClass)
+    {
+        assertNull(testClass.getEnclosingClass(), "Add support (and tests!) if you need it");
+        Class<?>[] interfaces = testClass.getInterfaces();
```
Comment (ebd2) [inline presto-connector-tests/src/main/java/com/facebook/presto/connector/meta/SupportedTestCondition.java]: Because putting the test methods in interfaces is apparently the JUnit 5 way and I haven't yet run into a scenario where I wanted or needed to put test methods in a class. I'm OK with adding it if we need it, but for now it seems like YAGNI unless you have a specific use in mind.

File: presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/BaseMetadataTest.java (hunk 4 above)
Comment (ebd2) [inline presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/BaseMetadataTest.java]: > Why do we need run method. Can't we just use:

Hive doesn't support some combinations of operations in the same transaction (adding a schema and listing schemas, IIRC).

> Still nicer looking approach would be to have withMetadata take multiple consumers:

https://github.com/prestodb/presto/pull/8672#discussion_r131481926

File: presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/BaseMetadataTest.java [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+        ConnectorSession session = new TestingConnectorSession(ImmutableList.of());
+
+        run(ImmutableList.of(
+                metadata -> assertEquals(metadata.listSchemaNames(session), systemSchemas()),
+                metadata -> assertEquals(metadata.listTables(session, null), ImmutableList.of())));
+    }
+
+    @Test
+    @RequiredFeatures({CREATE_TABLE, DROP_TABLE})
+    default void testCreateDropTable()
```
Comment (ebd2) [inline presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/BaseMetadataTest.java]: Commented.

File: presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/BaseMetadataTest.java [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+    default void withMetadata(Consumer<ConnectorMetadata> consumer)
+    {
+        Connector connector = getConnector();
+        ConnectorTransactionHandle transaction = connector.beginTransaction(IsolationLevel.READ_UNCOMMITTED, true);
+        ConnectorMetadata metadata = connector.getMetadata(transaction);
+        consumer.accept(metadata);
+        connector.commit(transaction);
+    }
+
+    class AllCloser
```
Comment (ebd2) [inline presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/BaseMetadataTest.java]: Didn't find Closer. Will change to it.

**Comment by ArturGajowy** [comment]
> Travis failure is expected. Code depends on unreleased airbase changes.

You could fix that by using jitpack.io. How you do this is you add jitpack maven repo and depend on com.github.aribase:airbase:commit_hash (or sth like that) and it builds the dependency for you on first attempt to resolve it. Just a hint, if you wanted the CI to provide feedback :)

File: presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/MetadataTableTest.java (hunk 5 above)
Comment (ebd2) [inline presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/MetadataTableTest.java]: API is now commented and more consistent, subject to the aforementioned constraint on variadic functions. Let me know if you think it's nicer ;-)

File: presto-postgresql/src/test/java/com/facebook/presto/unittests/TestPostgresqlMetadata.java (hunk 12 above)
Comment (ebd2) [inline presto-postgresql/src/test/java/com/facebook/presto/unittests/TestPostgresqlMetadata.java]: Commented.

In case it isn't clear from the comments:

> Shouldn't only schema related tests use the withSchemas method.

No. Anything that requires a table requires a schema. For e.g. Hive, you need to poof one into existence and then destroy it. For e.g. PostgreSQL, you already have it.

File: presto-hive/src/test/java/com/facebook/presto/hive/unittests/TestHiveMetadata.java (hunk 11 above)
Comment (ebd2) [inline presto-hive/src/test/java/com/facebook/presto/hive/unittests/TestHiveMetadata.java]: > Or just add getConnectorSpecificColumns method and to the concatenation of lists in tests?

That's how it was originally written. I changed it because this covers cases where the modification isn't just tacking some more columns on at the end of the list.

I've changed the name.

File: presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/BaseMetadataTest.java (hunk 6 above)
Comment (ebd2) [inline presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/BaseMetadataTest.java]: See below

File: presto-connector-tests/pom.xml (hunk 10 above)
Comment (ebd2) [inline presto-connector-tests/pom.xml]: Tabs/spaces. Fixed.

File: presto-connector-tests/src/main/java/com/facebook/presto/connector/meta/SupportedTestCondition.java (hunk 13 above)
Comment (ArturGajowy) [inline presto-connector-tests/src/main/java/com/facebook/presto/connector/meta/SupportedTestCondition.java]: It's great that there'll be an exception if s/b uses it contrary to that assumption, but:
1. if it's an assertion errror, the test will be reported as 'assertion failed' instead of 'error during test execution' - and it's the latter case here.
2. The exception message should tell explicitly that base classes are not supported (otherwise it's A+).

Comment (ArturGajowy) [inline presto-connector-tests/src/main/java/com/facebook/presto/connector/meta/SupportedTestCondition.java]: Now I see that the exception is for inner classes only and @losipiuk already pointed that out.

It also seems that there'll be no exception thrown or warning issued when `@RequiredFeatueres` is placed on a base class, which:
1. is confusing and will annoy users when they get to dig to the reason
2. seems against the javadoc

If there`s `@RequiredFeatures` on a base class in the hierarchy, an exception should be thrown telling it's not supported yet.

File: presto-generation-annotation/pom.xml [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>com.facebook.presto</groupId>
+        <artifactId>presto-root</artifactId>
+        <version>0.182-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>presto-generation-annotation</artifactId>
```
Comment (ArturGajowy) [inline presto-generation-annotation/pom.xml]: a stray file?

File: presto-connector-tests/src/main/java/com/facebook/presto/connector/meta/SupportedTestCondition.java (hunk 13 above)
Comment (ArturGajowy) [inline presto-connector-tests/src/main/java/com/facebook/presto/connector/meta/SupportedTestCondition.java]: You could also use gauva's [`TypeToken.getTypes()`](https://google.github.io/guava/releases/22.0/api/docs/com/google/common/reflect/TypeToken.html#getTypes--) and see that it's actually easier to implement the obvious expected functionality (base class requirements) than to not implement it properly (throwing an exception on misuse)

File: presto-connector-tests/src/test/java/com/facebook/presto/tpch/TestTpchMetadata.java [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+import com.google.common.collect.ImmutableMap;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.Iterables.getOnlyElement;
+
+/*
+ * Don't do this for other connectors. The TPCH connector is special because
```
Comment (losipiuk) [inline presto-connector-tests/src/test/java/com/facebook/presto/tpch/TestTpchMetadata.java]: Explicitly say what is `this`.
My understanding is `Don't add tests for other specific connectors to presto-connector-tests module`

File: presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/BaseMetadataTest.java [hunk 18]
```diff
@@ -0,0 +1,314 @@
...
+
+    /*
+     * Arguably, this belongs in MetadataTableTest. Unfortunately many
+     * connectors don't support CREATE_TABLE, but do support CREATE_TABLE_AS.
+     * As a result, MetadataTableTest is written in terms of CREATE_TABLE_AS,
+     * and is annotated with @RequiredFeatures({..., CREATE_TABLE_AS, ...) at
+     * the class level.
+     *
+     * Moving this to MetadataTableTest would require either:
+     * 1. Annotating every other methos than this as requiring CREATE_TABLE_AS.
```
[the remark this answers is not in the document]
Comment (ebd2) [reply presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/BaseMetadataTest.java]: If there's a consensus that annotating classes isn't needed, I'm just going to take all the code supporting that out of SupportedTestCondition. Please argue among yourselves and I'll either:

1. Take the code supporting annotations on classes out of STC, move this method, and update the RequiredFeatures on the existing methods
2. Implement the changes using TypeToken requested above and leave this here.

File: presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/BaseMetadataTest.java [hunk 19]
```diff
@@ -43,43 +44,121 @@
...
+     * (e.g. PostgreSQL with the "public" schema) should override this method
+     * and invoke callable without trying to create or destroy any of the
+     * schemas specified.
+     *
+     * Overriding this method in such a way allows tests that explicitly
+     * require creating and dropping tables to be written such that they are
+     * independent of whether or not the connector supports creating and
+     * dropping schemas.
+     */
+    default void withSchemas(ConnectorSession session, List<String> schemaNames, Callable<Void> callable)
```
[the remark this answers is not in the document]
Comment (maciejgrzybek) [reply presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/BaseMetadataTest.java]: Is an additional `return null` that bad compared to introducing an extra interface?
I'd probably prefer using standard interfaces as long as they fit (and an additional `return null` does not qualify for `not fit` to me).

[the remark this answers is not in the document]
Comment (ebd2) [reply presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/BaseMetadataTest.java]: I disagree that we should have withMetadata() methods with some number of Consumers<ConnectorMetadata>. It's a lot of boilerplate code to remove 3 instances of ImmutableList.of(). Which are lists of 2, 5, and 6 calls; should we implement the ones for 1, 3, and 4 calls in case we need them in the future?

I further disagree that we should do something like this with withSchemas and withTables. Most of the existing code that uses them is written in terms of the methods that take List<Consumer<Metadata>>. Should we have the cross product of number of schemas/tables and number of consumers just to save a couple ImmutableList.of() calls?

**Comment by ebd2** [comment]
Superseded by #8844

**Review by kokosing — COMMENTED** [review]
I haven't clicked the button in the past.
